// @oat-sa/tao-core-libs/src/pathdefinition.js
export default {
    jquery: '/node_modules/jquery/jquery',
    lodash: '/node_modules/lodash/lodash',
    moment: '/node_modules/moment/min/moment-with-locales',
    handlebars: '/node_modules/handlebars/dist/handlebars.amd',
    tpl: '/node_modules/@oat-sa/tao-core-libs/dist/tpl',
    'lib/dompurify/purify': '/node_modules/dompurify/dist/purify',
    'lib/moment-timezone.min': '/node_modules/moment-timezone/builds/moment-timezone-with-data.min',
    'lib/expr-eval/expr-eval': '/node_modules/@oat-sa/expr-eval/dist/bundle',
    async: '/node_modules/async/lib/async',
    'lib/decimal/decimal': '/node_modules/decimal.js/decimal',
    'lib/gamp/gamp': '/node_modules/gamp/src/gamp',
    interact: '/node_modules/interactjs/dist/interact',
    'lib/popper/tooltip': '/node_modules/tooltip.js/dist/umd/tooltip',
    popper: '/node_modules/popper.js/dist/umd/popper',
    raphael: '/node_modules/raphael/raphael',
    'select2-origin': '/node_modules/select2',

    // locals
    class: '/node_modules/@oat-sa/tao-core-libs/dist/class',
    'scale.raphael': '/node_modules/@oat-sa/tao-core-libs/dist/raphael/scale.raphael',
    nouislider: '/node_modules/@oat-sa/tao-core-libs/dist/sliders/jquery.nouislider',
    iframeNotifier: '/node_modules/@oat-sa/tao-core-libs/dist/iframeNotifier',
    lib: '/node_modules/@oat-sa/tao-core-libs/dist'
};
